### PR TITLE
Enhanced Rest Exceptions for AzConfig

### LINKED ISSUE #3323: Explore returning useful AzureServiceExceptions when a 4xx status is returned
It would be useful to deserialize the explanations and return them to the client callers.
For example, 4xx errors may have an explanation in the `WWW-Authenticate` header.

### PR DESCRIPTION
**Pull request by alzimmermsft**
Fixes #3323 

Additional exception types are thrown by the client based on the service response status code.

### CODE REVIEW

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java [hunk 1]
```diff
@@ -642,24 +643,32 @@ private static void assertConfigurationEquals(ConfigurationSetting expected, Con
...
-    private static void assertRestException(Throwable ex, int expectedStatusCode) {
-        assertTrue(ex instanceof ServiceRequestException);
-        assertEquals(expectedStatusCode, ((HttpRequestException) ex).response().statusCode());
+    private static void assertRestException(Runnable exceptionThrower, Class expectedExceptionType, int expectedStatusCode) {
+        try {
+            exceptionThrower.run();
+            fail();
+        } catch (Throwable ex) {
+            assertEquals(ex.getClass(), expectedExceptionType);
+            assertEquals(expectedStatusCode, ((HttpRequestException) ex).response().statusCode());
```
Comment (conniey) [inline applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: Is it possible to constrain the Class expectedExceptionType to be assignable to HttpRequestException? 🤔
You don't have to change it if it requires too much work. I was wondering about the cast here failing with an InvalidCastException if `ex` could not be casted to `HttpRequestException`.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 2]
```diff
@@ -198,9 +203,10 @@ public static ConfigurationClientBuilder builder() {
...
      * @throws IllegalArgumentException If {@link ConfigurationSetting#key() key} is {@code null}.
-     * @throws ServiceRequestException If a ConfigurationSetting with the same key and label does not
-     * exist, the setting is locked, {@link ConfigurationSetting#key() key} is an empty string, or
-     * {@link ConfigurationSetting#etag() etag} is specified but does not match the current value.
+     * @throws ResourceModifiedException If a ConfigurationSetting with the same key and label does not
+     * exist, the setting is locked, or {@link ConfigurationSetting#etag() etag} is specified but does not match
+     * the current value.
+     * @throws HttpRequestException If {@code key} is an empty string.
      */
     public Response<ConfigurationSetting> updateSetting(ConfigurationSetting setting) {
```
Comment (sima-zhu) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Just curious, do we need to throw ResourceNotFoundException when no key match?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationService.java [hunk 3]
```diff
@@ -33,32 +35,38 @@
...
     @ExpectedResponses({200})
+    @UnexpectedResponseExceptionType(code = {404}, value = ResourceNotFoundException.class)
     @UnexpectedResponseExceptionType(HttpRequestException.class)
     Mono<Response<ConfigurationSetting>> getKeyValue(@HostParam("url") String url, @PathParam("key") String key, @QueryParam("label") String label,
                                                      @QueryParam("$select") String fields, @HeaderParam("Accept-Datetime") String acceptDatetime,
                                                      @HeaderParam("If-Match") String ifMatch, @HeaderParam("If-None-Match") String ifNoneMatch);
 
     @PUT("kv/{key}")
     @ExpectedResponses({200})
+    @UnexpectedResponseExceptionType(code = {409, 412}, value = ResourceModifiedException.class)
```
Comment (sima-zhu) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationService.java]: 409 can be added into ResourceModifiedException Javadoc.

File: applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java (hunk 1 above)
Comment (alzimmermsft) [reply applicationconfig/client/src/test/java/com/azure/applicationconfig/ConfigurationClientTest.java]: Done

**Pull request merged**